src/System.IO.FileSystem.AccessControl/src/System/IO/FileSystemAclExtensions.netcoreapp.cs

```diff
@@ -24,5 +36,166 @@ public static void Create(this DirectoryInfo directoryInfo, DirectorySecurity di
+        {
+            if (fileInfo == null)
+            {
+                throw new ArgumentNullException(nameof(fileInfo));
+            }
+
+            if (fileSecurity == null)
+            {
+                throw new ArgumentNullException(nameof(fileSecurity));
+            }
+
+            // don't include inheritable in our bounds check for share
+            FileShare tempshare = share & ~FileShare.Inheritable;
+
+            string badArg = null;
+            if (mode < FileMode.CreateNew || mode > FileMode.Append)
+            {
+                badArg = nameof(mode);
+            }
+            else if (tempshare < FileShare.None || tempshare > (FileShare.ReadWrite | FileShare.Delete))
+            {
+                badArg = nameof(share);
+            }
+            if (badArg != null)
+            {
+                throw new ArgumentOutOfRangeException(badArg, SR.ArgumentOutOfRange_Enum);
+            }
+
+            // We need to check only the supported attributes
+            // NOTE: any change to FileOptions enum needs to be matched here in the error validation
+            if (options != FileOptions.None &&
+                (options & ~(FileOptions.WriteThrough | FileOptions.Asynchronous | FileOptions.RandomAccess | FileOptions.DeleteOnClose | FileOptions.SequentialScan | FileOptions.Encrypted | (FileOptions)0x20000000 /* NoBuffering */)) != 0)
+            {
+                throw new ArgumentOutOfRangeException(nameof(options), SR.ArgumentOutOfRange_Enum);
+            }
+
+            if (bufferSize <= 0)
+            {
+                throw new ArgumentOutOfRangeException(nameof(bufferSize), SR.ArgumentOutOfRange_NeedPosNum);
+            }
+
+            SafeFileHandle handle = null;
+            try
+            {
+                handle = CreateFileOpenHandle(fileInfo.FullName, mode, rights, share, options, fileSecurity);
+                return new FileStream(handle, GetFileStreamFileAccess(rights), bufferSize, (options & FileOptions.Asynchronous) != 0);
+            }
+            catch
+            {
+                // If anything goes wrong while setting up the stream, make sure we deterministically dispose of the opened handle.
+                if (handle != null)
+                {
+                    handle.Dispose();
+                    handle = null;
+                }
+
+                throw;
+            }
+        }
+
+        #endregion
+
+        #region Private methods
+
+        // In the context of a FileStream, the only ACCESS_MASK ACE rights we care about are reading/writing data and the generic read/write rights.
+        // See: https://docs.microsoft.com/en-us/windows/win32/secauthz/access-mask
+        private static FileAccess GetFileStreamFileAccess(FileSystemRights rights)
+        {
+            FileAccess access = 0;
+            if ((rights & FileSystemRights.ReadData) != 0 || ((int)rights & GENERIC_READ) != 0)
```

Comment:
I'm trying to follow, why are we supporting callers of the new `Create` method passing in a value of `rights` containing GENERIC_READ/GENERIC_WRITE? Why not only the documented entries in FileSystemRights ? Is it a compat thing I can't see?

Reply:
Hah, @JeremyKuhne was explaining me this earlier, and he mentioned how the GENERIC_* stuff is badly documented 😄 . So, my understanding from his explanations was that:

- Passing GENERIC_WRITE or GENERIC_READ is supposed to work for reading or writing, just like it would work to pass ReadData or WriteData.
- Generic rights are mutually exclusive with the specific rights (bits 0-15). The user should not pass both. <- I should add a test for this.

Reply:
It is a (simple) feature add essentially. We normally use `GENERIC_READ`/`GENERIC_WRITE` via `FileAccess`. We never defined them in `FileSystemRights`, but really should have. Doing this will allow people to specify `GENERIC_READ` & `GENERIC_WRITE` if they so choose and we won't fall over. It is technically safer to use that directly if that is what you want rather than select all of the flags it ends up being mapped to.

Reply:
Should we have an issue to add them to `FileSystemRights`?